unidata/ldmbridge/dcnldn/dcnldn.1
'\"
.TH NLDNTOGEM 1 "13 April 1994"
.SH NAME
\fInldntogem\fR - GEMPAK NLDN lightning decoder client for the Unidata LDM
.SH "UNIX SYNOPSIS"
\fBnldntogem [\fB\-l \fIlogfile\fR] \fR[\fB\-v\fR] \fR[\fB\-t timeout\fR]
\fIparm_file\fR   \fIfile_type\fR   \fIfile_suffix\fR   \fIfile_path\fR
.SH DESCRIPTION
This program decodes NLDN (National Lightning Detection Network) lightning 
reports broadcast from SUNY Albany and writes the decoded products into a 
GEMPAK surface data file.  The \fInldntogem\fR program can 
run as a subprocess of the LDM4 server or as a stand alone decoder. 

To run \fInldntogem\fR as a subprocess of the LDM server, it should be used
with the PIPE action from a pattern-action entry.  The form of such an entry 
is:

  NLDN	\fIpattern\fR	PIPE	\fInldntogem-executable\fR 
				\fIparm_file\fR \fIfile_type\fR \fIfile_suffix\fR
				\fIfile_path\fR

where \fIpattern\fR is a regular expression pattern matching the NLDN header.
The NLDN header currently consists of a two digit year, a three 
digit julian day (of the year), a two digit hour, a two digit minute, and
a two digit day (for example 94105120559).  This time represents the ending
time of the six minute period for which the lightning reports are sent.

The reports are decoded using the parameter packing file specified in
\fIparm_file\fR.  The reports are kept in monthly, daily, or hourly files
as specifed in \fIfile_type\fR (month, day, hour).  There is also a "minuteNN"
option for \fIfile_type\fR which uses an hourly file, but stores the data
inside the file in NN minute bins.  The default is 10 minute bins (minute10).  
The \fIfile_suffix\fR
argument determines the suffix used to name the file.  The decoded data
is stored in the directory specified by \fIfile_path\fR.  Any directories 
that do not exist in \fIfile_path\fR will be created as needed, so date 
components, for example, may be part of the output file path, specified by 
using matched parts of the regular expression \fIpattern\fR.

The filename itself is constructed by the decoder, based on the date and 
time of the individual lightning reports, the \fIfile_type\fR, and the 
\fIfile_suffix\fR.  For example, the following pattern-action 

.nf
NLDN ^(9[3-9])([0-3][0-9][0-9])([0-2][0-9])([0-5][0-9])([0-5][0-9])
	PIPE /home/gempak/bin/nldntogem 
	-l /var/data/ldm/gempak/nldn/log 
	/home/gempak/tables/nldn.pack 
	day   gem   /var/data/ldm/gempak/nldn/
.fi

will result in decoding all the data for April 15, 1994 into the daily file 
940415.gem.

The \fInldntogem\fR executable must be specified as the first argument to 
the PIPE command, and should be either an absolute path name 
or in a directory in the executable search path of the LDM4 server.

The \fInldntogem\fR program may also be used as a stand alone decoder
to process data read from standard input.  For example, if raw NLDN
lightning reports for 1200-1206Z April 15, 1994 are stored in the 
file "nldnraw":

.nf
nldntogem -l log /home/gempak/tables/nldn.pack hour nldn ./ < nldnraw
.fi

will create a GEMPAK data file 94041512.nldn in the current directory
and decode the raw reports in "nldnraw" into it.

The GEMPAK data file created by \fInldntogem\fR is a SHIP-type surface
file, where each report contains the date, time, and location of the
flash along with the requested data values.  Available data include:
TSEC (Seconds since STIM (station time)), MSEC (Milliseconds since TSEC),
SGNL (Signal strength and polarity where 150 units corresponds to 
approximately 30 KAmps), MULT (Multiplicity in strokes/flash), SMAJ 
(Semi-major axis of ellipse), ECNT (Eccentricity 
of ellipse), ANGL (Angle in degrees), and CHI2 (Chi-squared confidence 
measure).  Which of the data values are decoded is determined by the
parameter packing file \fIparm_file\fR.  

The GEMPAK data file will
hold up to 19,999 reports.  If 20,000 or more reports are sent to
a single data file, additonal files are created by appending a
".more" to the file name.  In the LDM example above, flashes 20,000-
39,999 will be stored in file 940415.gem.more.  Flashes 40,000-59,999
will be stored in file 940415.gem.more.more, etc.  NLDN has reported that
up to 35,000 flashes have been detected in A SINGLE HOUR during 
peak lightning season.  


.SH OPTIONS 
.TP 8
.BI \-l " logfile"
The path and  name of a file to be used as the log file for the process.  
If "-" is specified as the logfile or no logfile is specified, the default is
standard error.
.TP
.BI \-v
Verbose flag.  If present, the log will receive more detailed messages
about the NLDN decoding, including summary of totals decoded and written.
.TP 8
.BI \-t " timeout"
Timeout in seconds (default is 600).  If there has been no input available
on stdin for the duration of the specified timeout, the program exits
gracefully.
.TP
.BI "parm_file"
The GEMPAK parameter packing file listing the data values to decode and store in 
GEMPAK file.  Usually /home/gempak/tables/nldn.pack.
.TP
.BI "file_type"
(hour day month minuteNN) Specifies whether hourly, daily, or monthly files are
to be created by the decoder.  The minuteNN option keeps hourly files, but the
reports are stored in NN minute bins.  For example minute06 will generate
a single hour file with YYMMDD/HH00, YYMMDD/HH06, YYMMDD/HH12, ... times.
Please note that the limit on number of lightning
reports in a single file is 19,999.  During peak lightning season, up to
35,000 flashes have been seen in a single hour.  Because \fInldntogem\fR will
open additional data files if necessary (by appending a ".more" to the file
name), you will not lose any reports if the number of flashes is
extraordinarily high.  However, you may find that appropriate values for 
\fIfile_type\fR (day vs. hour) may change with the season.
.TP
.BI "file_suffix"
This parameter is used to generate the name given to the GEMPAK data file.
You may use any suffix you like.  Examples include "gem", "nldn", and 
"lightning", leading to filenames such as 940412.gem, 940412.nldn, and
940412.lightning.
.TP
.BI "file_path"
The path name of the directory that will contain the GEMPAK data file.
This may be specified as an absolute path name or (more typically) relative 
to the directory from which \fInldntogem\fR is invoked.  Directories that 
do not exist in the path name of the output file will be created as needed.
NOTE:  THE \fIfile_path\fR MUST CONTAIN A TRAILING "/".  For example:
"/usr/local/gempak/data/nldn/".



.SH
Finally, the \fInldntogem\fR GEMPAK lightning data decoder and the data
that results from its use has many subtleties associated with it.  It is
strongly reccommended that you carefully read the nldntogem.README file
that is supplied along with the \fInldntogem\fR decoder software.


.SH "SEE ALSO"
satogem(1), ldmd(1)
.SH DIAGNOSTICS
Error messages and log messages are written to the log file.



unidata/ldmbridge/dcnldn/dcnldn.c
/*
** Do not change these include commands.
** All necessary header files are included in "geminc.h".
** All macros and constants are in "gemprm.h" and "bridge.h".
*/
#include "geminc.h"
#include "gemprm.h"
#include "bridge.h"
#include "dcnldn.h"

int main ( int argc, char *argv[] )
/************************************************************************
 * DCNLDN								*
 *									*
 * This program decodes NLDN lightning reports and stores them in a	*
 *   GEMPAK ship file. 			                                *
 *									*
 *  Command line:							*
 *  dcnldn [options] filename						*
 *      filename        output file name/template                       *
 **									*
 * Log:									*
 * Chiz/Unidata		8/96	Created from nldntogem 			*
 ***********************************************************************/
{
/*
**      Change the values of these default variables for the
**      specific decoder.
**
**      These variables are the number of expected command line
**      parameters; the program name; the packing and station tables;
**      values for the the number of additional stations and the
**      number of times; and the number of hours, prior to the
**      "current" time, to decode.
*/
#define NUMEXP	1
	int	nexp    = NUMEXP;
	char	*prgnam = "DCNLDN";

	char	*defprm = "nldn.pack";
	char	*defstn = "hour";
	char	*dfstn2 = " ";
	int	idfadd  = 0;
	int	idfmax  = 25000;
	int	ndfhr1  = 60; /* default -b file NN bins */
	int	ndfhr2  = 24;
	int	idfwdh  = 0;

/*
**	Do not change these variables. These variables are used by all
**	decoders for getting the command line parameters.
*/
	char	parms[NUMEXP][DCMXLN], curtim[DCMXLN];
	int	num, iret;

	char	gemfil[DCMXLN], stntbl[DCMXLN], stntb2[DCMXLN],
		prmfil[DCMXLN];
	int	iadstn, maxtim, nhours, txtflg, crcflg, iwndht;

/*---------------------------------------------------------------------*/

/*
**	Initialize the output logs, set the time out and 
**	parse the command line parameters. dc_init calls in_bdta in 5.4.3+. 
*/
	dc_init ( prgnam, argc, argv, nexp, parms, &num, &iret );

/*
**	Check for an initialization error.
**	On an error, exit gracefully.
*/
	if  ( iret < 0 )  {
	    dc_exit ( &iret );
	}

/*
**	Set the decoder parameters to the command line entries or
**	default values.
*/
	dc_gopt ( defprm, defstn, dfstn2, idfadd, idfmax, ndfhr1, ndfhr2,
		  idfwdh, prmfil, stntbl, stntb2, &iadstn, &maxtim, curtim, &nhours,
		  &txtflg, &crcflg, &iwndht, &iret );

/*
**	The output file name must be present.
**
**	Change this section for the specific decoder.
*/
	strcpy ( gemfil, parms[0] );
        /* add %% template for file sequence if not @@ */
        if(strstr(gemfil,"@@") == NULL)
           strncat(gemfil,"_%%",3);

/*
**	Call the decoding routine.
**
**	Change this function call, and the define command,
**	for the specific decoder.
*/
	decode_nldn (curtim, gemfil, prmfil, stntbl, maxtim,
		nhours, &iret);

/*
**	Send shut down message and close the log files.
*/
	dc_exit ( &iret );

}
